yapapi/engine.py

```diff
@@ -501,52 +501,24 @@ async def _enforce_debit_note_intervals(self, job: "Job", debit_note: DebitNote)
 
     async def _process_debit_notes(self) -> None:
         """Process incoming debit notes."""
+        max_concurrent_notes = 10
+        debit_note_processing_tasks: Dict[str, asyncio.Task] = {}
+        loop = asyncio.get_event_loop()
 
-        async for debit_note in self._payment_api.incoming_debit_notes():
-            agr_id = debit_note.agreement_id
-            job_id = next(
-                (
-                    id
-                    for id in self._agreements_accepting_debit_notes
-                    if agr_id in self._agreements_accepting_debit_notes[id]
-                ),
-                None,
-            )
-            if job_id is not None:
-                job = self._get_job_by_id(job_id)
-                agreement = self._get_agreement_by_id(agr_id)
-                job.emit(
-                    events.DebitNoteReceived,
-                    agreement=agreement,
-                    debit_note=debit_note,
-                )
-
-                # We ignore debit notes we can't accept, since rejection is not implemented in yagna
-
-                # The most we risk by not accepting a debit note would be a termination of the
-                # agreement by the provider which is not an issue here
-                # because in all of these cases the agreement had already been terminated or
-                # we have just terminated it in the course of interval enforcement
-                if not await self._enforce_debit_note_intervals(job, debit_note):
-                    continue
+        semaphore = asyncio.Semaphore(max_concurrent_notes)
 
-                try:
-                    allocation = self._get_allocation(debit_note)
-                    await debit_note.accept(
-                        amount=debit_note.total_amount_due, allocation=allocation
-                    )
-                    job.emit(
-                        events.DebitNoteAccepted,
-                        agreement=agreement,
-                        debit_note=debit_note,
-                    )
-                except CancelledError:
-                    raise
-                except Exception:
-                    job.emit(
-                        events.PaymentFailed, agreement=agreement, exc_info=sys.exc_info()  # type: ignore
-                    )
+        async def _process_debit_note_wrapper(debit_note_id: str) -> None:
+            try:
+                await self._process_debit_note(debit_note_id)
+            finally:
+                debit_note_processing_tasks.pop(debit_note_id, None)
+                semaphore.release()
 
+        async for debit_note_id in self._payment_api.incoming_debit_note_ids():
+            await semaphore.acquire()
```

Comment:
is there a reason why we're doing this and not:


```
async with semaphore:
    try:
        self._process_debit_note(debit_note_id)
    finally:
        debit_note_processing_tasks.pop(debit_note_id, None)
```

Reply:
Good idea, `async with` is much better than manual handling of this `Semaphore`. I used the code from `_find_offers_for_subscription` as I knew it works well.

Reply:
Now I'm not so sure. We cannot put `async with semaphore` inside `_process_debit_note_wrapper` launched by `loop.create_task` as this would not limit the number of tasks created. I think the semaphore has to be released in a different task/coroutine than it is `acquire`d.

Reply:
okay, but we can create the task and await it inside the `async with`

Reply:
that's a pretty good point